File: medium/_0807_Max IncreasetoKeepCitySkyline.py
# 807. Max Increase to Keep City Skyline

"""https://leetcode.com/problems/max-increase-to-keep-city-skyline/discuss/120681/C%2B%2BJavaPython-Easy-and-Concise-Solution"""

class Solution(object):
    def maxIncreaseKeepingSkyline(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        row, col = map(max, grid), map(max, zip(*grid))
        return sum(min(i, j) for i in row for j in col) - sum(map(sum, grid))

File: medium/_0807_Max IncreasetoKeepCitySkyline.cpp
// 807. Max Increase to Keep City Skyline

class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        auto n = grid.size();
		vector<int> viewed_top(n);
		vector<int> viewed_left(n);
		
		for (size_t i = 0; i < n; i++)
		{
			for (size_t j = 0; j < n; j++)
			{
				viewed_top[i] = max(viewed_top[i], grid[i][j]);
				viewed_left[j] = max(viewed_left[j], grid[i][j]);
			}
		}

		int output = {};
		for (size_t i = 0; i < n; i++)
		{
			for (size_t j = 0; j < n; j++)
			{
				output += min(viewed_top[i], viewed_left[j]) - grid[i][j];
			}
		}

		return output;
    }
};